lib/grid.h

```diff
@@ -10,178 +10,941 @@
+          return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Check if a multidimensional index is within grid bounds
+   * @param gidx grid index
+   * @return true if within bounds
+   */
+  bool Valid(const Vectorni& gidx) const{
+    for(size_t i=0; i< n; i++){
+      if(gidx[i]<0 || gidx[i]>=gs[i])
         return false;
     }
     return true;
   }
 
   /**
-   * Get an id of point x useful for direct lookup in the grid array
-   * @param x point
-   * @return a computed id
+   * Check if id( array lookup id) is valid
+   * @param id fast lookup id on array
+   * @return true if within bounds
    */
-  int Id(const PointType& x) const {
+  bool Valid(int id) const{
+    return (id>=0 && id<nc);
+  }
 
-       // unroll loop for n=1,2,3,4 for efficiency
-    if (n==1)
-      return Index(x, 0);
+  /**
+   * Finds the CellCenter corresponding to the input idx
+   * @param x The center point (meaningful even when gidx outside bounds)
+   * @param gidx Multidimensional index (can extend outside the grid)
+   * @return true if cell center inside grid bounds
+   */
+  bool CellCenter(Vectornd& x,const Vectorni& gidx) const{
+    x = xlb.array() +  (gidx.template cast<double>() + Vectornd::Constant(0.5)).array()*cs.array() ;
+    return Valid(gidx);
+  }
+
+  /**
+   * Finds the CellCenter corresponding to the input idx
+   * @param x The updated center ( set to a vector of NANs when id not valid)
+   * @param id fast access id in the cell array
+   * @return true if id is valid else false
+   */
+  bool CellCenter(Vectornd& x, int id) const{
+    Vectorni gidx;
+    if(!Index(gidx,id)){
+      x = Vectornd::Constant(numeric_limits<double>::quiet_NaN());
+      return false;
+    }
+    x = xlb.array() +  (gidx.template cast<double>()+Vectornd::Constant(0.5)).array()*cs.array() ;
+    return true;
+  }
 
-    if (n==2)
-      return Index(x, 0) + gs[0]*Index(x, 1);
+  /**
+   * Returns cc[i] where cc is the cell center of any cell with gidx[i] = idx
+   * @param idx gidx[i] = idx, i.e. grid index along dimension i.
+   * @param i coordinated index/ dimension
+   * @return Returns cc[i]. Meaningful result even when idx out of range
+   */
+  double CellCenterIth(int idx, int i) const{
+    return xlb[i] + (idx +0.5)*cs[i];
+  }
+
+  /**
+   * Get id of point x useful for direct lookup in the cell array
+   * @param x point
+   * @return the cell array id. -1 if point x is not in grid
+   */
+  int Id(const Vectornd& x) const {
+    Vectorni gidx; Index(gidx,x);
+    if(Valid(gidx))
+      return gidx.transpose()*cgs;
+    else
+       return -1;
+  }
 
-    if (n==3)
-      return Index(x, 0) + gs[0]*Index(x, 1) + gs[0]*gs[1]*Index(x, 2);
+  /**
+   * Get the id of a point corresponding to the grid index
+   * @param gidx grid index
+   * @return the cell array id. -1 if gidx is not within bounds
+   */
+  int Id(const Vectorni& gidx) const {
+    if(Valid(gidx))
+      return gidx.transpose()*cgs;
```

Comment:
I benchmarked against the old code(with checks for out of bound) and found the following elapsed times for 1e7 calls:

grid2.gs():10 10
dt Id:0.113313 secs
dt IdOld:0.0857138 secs

grid3.gs():10 10 10
dt Id:0.170228 secs
dt IdOld:0.127534 secs

grid4.gs():10 10 10 10
dt Id:0.220866 secs
dt IdOld:0.16996 secs

grid5.gs():10 10 10 10 10
dt Id:0.230863 secs
dt IdOld:0.212158 secs

Also gidx.transpose()*cgs is faster than gidx.dot(cgs)